### DISCUSSION TITLE
Plotting streamlines

### DISCUSSION THREAD

**Discussion by SantiagoPendon** [pyvista/pyvista#5234] [Q&A]
Hi to everyone. On this time I'm trying to plot streamlines from CFD data. The process I follow is: https://docs.pyvista.org/version/stable/examples/01-filter/streamlines.html#carotid , example 2 'blood vessels'

The problem is when I compute the  streamlines because are empty.  The procedure is quite simple and exactly the next one:
```py
velocity = np.vstack((df_velocity['Velocity[i] (m/s)'].to_numpy(), df_velocity['Velocity[j] (m/s)'].to_numpy(), df_velocity['Velocity[k] (m/s)'].to_numpy())).T.astype(np.float32)

pos_velocity = np.vstack((df_velocity['X (m)'].to_numpy(), df_velocity['Y (m)'].to_numpy(), df_velocity['Z (m)'].to_numpy())).T*1000

vel_poly = pv.PolyData(pos_velocity)
vel_poly.point_data['velocity'] = velocity
vel_grid = pv.UnstructuredGrid(vel_poly)


vel_grid.set_active_scalars("velocity")

streamlines, src = vel_grid.streamlines(return_source=True,source_center=(0, 33, -163))

sargs = dict(vertical=True, title_font_size=16)
p = pv.Plotter()
p.add_mesh(streamlines.tube(radius=0.2), lighting=False, opacity=0.25)
p.add_mesh(src)
p.show()


vel_grid is:
UnstructuredGrid (0x1ee1f08c8e0)
  N Cells:    663362
  N Points:   663362
  X Bounds:   -8.844e+01, 2.609e+01
  Y Bounds:   -1.474e+01, 9.223e+01
  Z Bounds:   -2.364e+02, -1.368e+02
  N Arrays:   1

and vel_grid['velocity']:
pyvista_ndarray([[-0.01877574,  0.00661109, -0.01557323],
                 [-0.00915081,  0.002352  , -0.0080691 ],
                 [-0.02685505,  0.00925507, -0.01925328],
                 ...,
                 [ 0.        ,  0.        ,  0.        ],
                 [ 0.        ,  0.        ,  0.        ],
                 [ 0.        ,  0.        ,  0.        ]], dtype=float32)

and streamlines:
PolyData (0x1ee1f08e020)
  N Cells:    0
  N Points:   0
  N Strips:   0
  X Bounds:   1.000e+299, -1.000e+299
  Y Bounds:   1.000e+299, -1.000e+299
  Z Bounds:   1.000e+299, -1.000e+299
  N Arrays:   5
```
My pyVista report is:

![Captura de pantalla 2023-11-22 083818](https://github.com/pyvista/pyvista/assets/151015554/124e33c7-991a-4a5a-9f44-454469f8f9fe)

**Comment by MatthewFlamm** [comment]
I cannot run your code, but the provided mesh information is helpful.  Thank you for including that.  You start with `PolyData` and only provide points.  This means each point is a vertex cell, i.e. there is no topological structure between the points. Or, another way, they are 0D cells.  The `vel_grid` has the same number of points and cells, which demonstrates this.  streamlines are generated inside a mesh with 2D or 3D cell structures.  Presuming you are using a typical CFD technique with a cell based mesh, e.g. FVM, then you should use the mesh topology of that mesh.

**Reply by SantiagoPendon** [reply]
Hi Matthew. Thanks for your answer, but i don't see the point. 

I have two meshes:

- The first one is provided by the .stl and has 200k points
- The second one is provided by CFD simulation and is solid (point cloud) and has 600k points. On this points is where I have velocity info which I try to streamline, but I can't. Note that I convert it into mesh, but in the beginings are only points.

What I suppose to do?

**Reply by MatthewFlamm** [reply]
Can you share how you get the CFD data?  Is your mesh unstructured or structured in the simulation?  Or is it a meshless approach like SPH?

**Reply by SantiagoPendon** [reply]
![Captura de pantalla 2023-11-27 140911](https://github.com/pyvista/pyvista/assets/151015554/fcfdb0de-6b82-463f-b16b-34f247530e7a)

This the data from the point cloud. The simulation runs in structured mesh

**Reply by MatthewFlamm** [reply]
The structure isn't so obvious to me. 
You would use `UniformGrid` if your mesh has rectangular cells aligned to the cartesian axes and each cell has the same size, `RectilinearGrid` if the mesh has rectangular cells aligned to the cartesian axes but cells are different sizes, or `StructuredGrid` if there is a structured i-j-k cell ordering but the cells could be any arbitrary hexahedral shape (including degenerate ones).
